eth_abi/registry.py

```diff
@@ -0,0 +1,146 @@
+from collections import namedtuple
+
+
+class ABIRegistryError(Exception):
+    pass
+
+
+class ExistingEntryFound(ABIRegistryError):
+    pass
+
+
+class NoEntryFound(ABIRegistryError):
+    pass
+
+
+PredicateItem = namedtuple('PredicateItem', ('predicate', 'value'))
+
+
+def ensure_changed(ExcClass, error_msg):
+    def decorator(old_method):
+        def new_method(self, *args, **kwargs):
+            len_before = len(self._items)
+            result = old_method(self, *args, **kwargs)
+            len_after = len(self._items)
+
+            if len_before == len_after:
+                raise ExcClass(error_msg)
+
+            return result
+
+        return new_method
+
+    return decorator
+
+
+class PredicateList:
+    def __init__(self):
+        self._items = []
+
+    def add(self, predicate, value):
+        self._items.insert(0, PredicateItem(predicate, value))
+
+    def match(self, *args, **kwargs):
+        return tuple(i.value for i in self._items if i.predicate(*args, **kwargs))
+
+    @ensure_changed(KeyError, 'Given predicate not found')
+    def remove_exact(self, predicate):
+        self._items = [i for i in self._items if i.predicate is not predicate]
+
+    @ensure_changed(KeyError, 'No matching items found')
+    def remove_matching(self, *args, **kwargs):
```

Comment:
This doesn't appear to be used.

Reply:
Yep, things aren't done yet so I just have that in there in case I want to unregister callbacks in a different way.